sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala

```diff
@@ -261,7 +261,7 @@ final class Decimal extends Ordered[Decimal] with Serializable {
 
   def - (that: Decimal): Decimal = Decimal(toBigDecimal - that.toBigDecimal)
 
-  def * (that: Decimal): Decimal = Decimal(toBigDecimal * that.toBigDecimal)
+  def * (that: Decimal): Decimal = Decimal(toJavaBigDecimal.multiply(that.toJavaBigDecimal))
```

Comment:
I think this can be fixed by in `toBigDecimal`:
```
BigDecimal(longVal, _scale, new MathContext(precision, RoundingMode.HALF_EVEN))
```

Also, in order to have the same behavior as other datebase or Hive, we should throw an exception if precision is higher than 38.